### ISSUE TITLE
type inference bug with nested closures

### ISSUE THREAD

**Issue by mraleph** [dart-lang/kernel#24]
In the code below `X._x` is inferred to be `Null` which is incorrect. If line `// _x = data;` is uncommented then inference is correct - which means something is going wrong with connecting captured variables.  

```dart
class X {
  var _x;

  foo() {
    bar((data) {
      baz(() {
        _x = data;
      });
      // _x = data; 
    });
  }

  bar(f) {
    f(10);
    f(new X());
  }

  baz(f) => f();
}

main() {
  var obj = new X();
  obj.foo();
}
```

In fact even something like:

```dart
  foo() {
    var x = qux();
    f() {
      _x = x;
    }
    f();
  }

  qux() {
    return "a";
  }
```

yield incorrect results - which indicates that there is an issue with captured variables even without nested closures.

**Comment by mraleph** [comment]
Yeah `Environment.getVariable` does not respect environment chain in any way... It treats all variables as local variables - I am surprised that we did not notice it before. My fault.

**Comment by mraleph** [comment]
I uploaded a tentative fix for this issue: https://chromereviews.googleplex.com/489217013/

**Comment by asgerf** [comment]
Dang. I remember fixing this bug before, but I must have forgotten to cherry pick the fix after the branch I worked on died.

I put up my original fix here https://chromereviews.googleplex.com/492317013.

**Comment by asgerf** [comment]
Fixed in e7cae46b5276995d3246d0616d9aebbaa8bb2065
